src/array/cpu/coo_union_partition.cc

```diff
@@ -0,0 +1,122 @@
+/*!
+ *  Copyright (c) 2020 by Contributors
+ * \file array/cpu/coo_union_partition.cc
+ * \brief COO union and partition
+ */
+#include <dgl/array.h>
+#include <vector>
+
+namespace dgl {
+namespace aten {
+namespace impl {
+
+template <DLDeviceType XPU, typename IdType>
+COOMatrix DisjointUnionCooGraph(const std::vector<COOMatrix>& coos,
+                                const std::vector<uint64_t> src_offset,
+                                const std::vector<uint64_t> dst_offset) {
+  std::vector<int64_t> edge_offset(coos.size());
+  int64_t total_edges = 0;
+  bool has_data = false;
+
+  for (size_t i = 0; i < coos.size(); ++i) {
+      aten::COOMatrix coo = coos[i];
+      edge_offset[i] = total_edges;
+      total_edges += coo.row->shape[0];
+
+      if (IsNullArray(coo.data) == false)
+        has_data = true;
+  }
+
+  std::vector<IdType> result_src(total_edges);
+  std::vector<IdType> result_dst(total_edges);
+  std::vector<IdType> result_data;
+  if (has_data)
+    result_data.resize(total_edges);
+
+#pragma omp parallel for
+  for (int64_t i = 0; i < coos.size(); ++i) {
+    aten::COOMatrix coo = coos[i];
+    int64_t num_edges = coo.row->shape[0];
+
+    const IdType* edges_src_data = static_cast<const IdType*>(coo.row->data);
+    const IdType* edges_dst_data = static_cast<const IdType*>(coo.col->data);
+    const IdType* data = static_cast<const IdType*>(coo.data->data);
+
+    // Loop over all edges
+    for (int64_t j = 0; j < num_edges; ++j) {
+      result_src[edge_offset[i] + j] = edges_src_data[j] + src_offset[i];
+      result_dst[edge_offset[i] + j] = edges_dst_data[j] + dst_offset[i];
+
+      if (has_data) {
+        if (data == nullptr) {
+          result_data[edge_offset[i] + j] = edge_offset[i] + j;
+        } else {
+          result_data[edge_offset[i] + j] = edge_offset[i] + data[j];
+        }
+      }
+    }
+  }
```

Comment:
Suggest use array operators to rewrite this. Pseudo-codes:
```python
rows = []
cols = []
nrows = 0
ncols = 0
for coo in coos:
    rows.append(coo.row + nrows)
    cols.append(coo.col + ncols)
new_row = concat(rows)
new_cols = concat(cols)
...
```

This means the implementation is actually device- and idtype-agnostic and you can directly implement it in `array.cc` without creating separate implementations for CPU and GPU.